aquadoggo/src/replication/service.rs

```diff
@@ -0,0 +1,398 @@
+
+use super::SessionId;
+
+const MAX_SESSIONS_PER_CONNECTION: usize = 3;
+
+#[derive(PartialEq, Eq, PartialOrd, Clone, Debug, Hash)]
+
+/// Identifier for a connection to another peer. The `ConnectionId` is optional as we need to
+/// identify the local peer in the replication manager, but in this case there is no single
+/// connection to associate with.
+///
+/// @TODO: This could be modelled better maybe...
+pub struct PeerConnectionId(PeerId, Option<ConnectionId>);
+
+impl Display for PeerConnectionId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}{:?}",
+            self.0,
+            // If we don't pass a connection id then we default to `0`
+            self.1.unwrap_or(ConnectionId::new_unchecked(0))
+        )
+    }
+}
+
+pub async fn replication_service(
+    context: Context,
+    shutdown: Shutdown,
+    tx: ServiceSender,
+    tx_ready: ServiceReadySender,
+) -> Result<()> {
+    // Subscribe to communication bus
+    let _rx = tx.subscribe();
+
+    let local_peer_id = context
+        .config
+        .network
+        .peer_id
+        .expect("Peer id needs to be given");
+
+    // Define set of schema ids we are interested in
+    let supported_schema_ids: Vec<SchemaId> = context
+        .schema_provider
+        .all()
+        .await
+        .iter()
+        .map(|schema| schema.id().to_owned())
+        .collect();
+    let target_set = TargetSet::new(&supported_schema_ids);
+
+    // Run a connection manager which deals with the replication logic
+    let manager = ConnectionManager::new(
+        &context.schema_provider,
+        &context.store,
+        &tx,
+        local_peer_id,
+        target_set,
+    );
+
+    let scheduler_handle = task::spawn(ConnectionManager::start_scheduler(tx.clone()));
+    let handle = task::spawn(manager.run());
+
+    if tx_ready.send(()).is_err() {
+        warn!("No subscriber informed about replication service being ready");
+    };
+
+    tokio::select! {
+        _ = handle => (),
+        _ = scheduler_handle => (),
+        _ = shutdown => {
+            // @TODO: Wait until all pending replication processes are completed during graceful
+            // shutdown
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct PeerStatus {
+    peer_id: PeerId,
+    successful_count: usize,
+    failed_count: usize,
+}
+
+impl PeerStatus {
+    pub fn new(peer_id: &PeerId) -> Self {
+        Self {
+            peer_id: *peer_id,
+            successful_count: 0,
+            failed_count: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct PeerConnections {
+    status: PeerStatus,
+    connections: Vec<ConnectionId>,
+}
+
+struct ConnectionManager {
+    connections: HashMap<PeerId, PeerConnections>,
+    sync_manager: SyncManager<PeerConnectionId>,
+    tx: ServiceSender,
+    rx: Receiver<ServiceMessage>,
+    target_set: TargetSet,
+}
```

Comment:
I extended the map of peers to include all connections to each peer (copied from libp2p gossip protocol implementation).